File: buildSrc/src/main/java/com/mars/infra/router/plugin/RouterTransform.java
package com.mars.infra.router.plugin;

import com.android.build.api.transform.DirectoryInput;
import com.android.build.api.transform.Format;
import com.android.build.api.transform.JarInput;
import com.android.build.api.transform.QualifiedContent;
import com.android.build.api.transform.Transform;
import com.android.build.api.transform.TransformException;
import com.android.build.api.transform.TransformInput;
import com.android.build.api.transform.TransformInvocation;
import com.android.build.api.transform.TransformOutputProvider;
import com.android.build.gradle.internal.pipeline.TransformManager;
import com.android.utils.FileUtils;
import com.mars.infra.router.plugin.visitor.RouterClassVisitor;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Opcodes;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.jar.JarOutputStream;
import java.util.zip.ZipEntry;

/**
 * Created by JohnnySwordMan on 2022/1/8
 */
class RouterTransform extends Transform {

    // 所有注册类
//    private static List<String> registerList = new ArrayList<>();
    // Router.class所在的jar包
//    private static File destFile;

    @Override
    public String getName() {
        return RouterTransform.class.getSimpleName();
    }

    /**
     * 处理类型
     */
    @Override
    public Set<QualifiedContent.ContentType> getInputTypes() {
        return TransformManager.CONTENT_CLASS;
    }

    /**
     * 处理范围
     */
    @Override
    public Set<? super QualifiedContent.Scope> getScopes() {
        return TransformManager.SCOPE_FULL_PROJECT;
    }

    @Override
    public boolean isIncremental() {
        return false;
    }

    @Override
    public void transform(TransformInvocation transformInvocation) throws TransformException, InterruptedException, IOException {
        super.transform(transformInvocation);

        if (transformInvocation == null) {
            return;
        }

        if (!transformInvocation.isIncremental()) {
            transformInvocation.getOutputProvider().deleteAll();
        }

        Collection<TransformInput> inputs = transformInvocation.getInputs();
        TransformOutputProvider outputProvider = transformInvocation.getOutputProvider();
        for (TransformInput input : inputs) {
            // apply该插件的模块的所有的类以directory的形式获取
            Collection<DirectoryInput> directoryInputs = input.getDirectoryInputs();
            handleDirectory(directoryInputs, outputProvider);
            // 本地依赖或者远程依赖，以jar包的形式获取
            Collection<JarInput> jarInputs = input.getJarInputs();
            handleJar(jarInputs, outputProvider);
        }
        // RouterTransform all mapping className = [UriAnnotationInit_996, UriAnnotationInit_2677]
        System.out.println("RouterTransform all mapping className = " + RouterMappingCollector.getRouterMapping());
        if (RouterMappingCollector.getDestFileOfRouter() != null) {
//            RegisterCodeGenerator.insertInitCode(registerList, destFile);
//            RegisterCodeGenerator.insertInitCode(RouterMappingCollector.getRouterMapping(), RouterMappingCollector.getDestFileOfRouter());
        }

//        File mappingJarFile = outputProvider.getContentLocation(
//                "router_mapping",
//                getOutputTypes(),
//                getScopes(),
//                Format.JAR);
//        System.out.println("RouterTransform mappingJarFile = " + mappingJarFile.getAbsolutePath());
//        if (mappingJarFile.getParentFile().exists()) {
//            mappingJarFile.getParentFile().mkdirs();
//        }
//        if (mappingJarFile.exists()) {
//            mappingJarFile.delete();
//        }

        // 将mappingJarFile写入本地
//        FileOutputStream fileOutputStream = new FileOutputStream(mappingJarFile);
//        JarOutputStream jarOutputStream = new JarOutputStream(fileOutputStream);
//        ZipEntry zipEntry = new ZipEntry(RouterMappingByteCodeBuilder.CLASS_NAME + ".class");
//        jarOutputStream.putNextEntry(zipEntry);
//        jarOutputStream.write(RouterMappingByteCodeBuilder.get(RouterMappingCollector.getRouterMapping()));
//        jarOutputStream.closeEntry();
//        jarOutputStream.close();
//        fileOutputStream.close();
    }


    /**
     * 遍历class文件，找到com.mars.infra.router.UriAnnotationInit_996
     */
    private void handleDirectory(Collection<DirectoryInput> directoryInputs, TransformOutputProvider outputProvider) throws IOException {
        for (DirectoryInput input : directoryInputs) {
            File destFile = outputProvider.getContentLocation(
                    input.getFile().getAbsolutePath(),
                    input.getContentTypes(),
                    input.getScopes(),
                    Format.DIRECTORY);
//            RouterMappingCollector.collectFromDirectory(input.getFile());
            FileUtils.copyFile(input.getFile(), destFile);
        }
    }

    private void handleJar(Collection<JarInput> jarInputs, TransformOutputProvider outputProvider) throws IOException {
        for (JarInput input : jarInputs) {
            File destFile = outputProvider.getContentLocation(
                    input.getFile().getAbsolutePath(),
                    input.getContentTypes(),
                    input.getScopes(),
                    Format.JAR);
//            RouterMappingCollector.collectFromJarFile(input.getFile());
            FileUtils.copyFile(input.getFile(), destFile);
        }
    }

    private void processDir(File input, File dest) throws IOException {
        if (input == null || dest == null) {
            return;
        }
        if (dest.exists()) {
            FileUtils.delete(dest);
        }
        FileUtils.mkdirs(dest);
        String srcDirPath = input.getAbsolutePath();
        String destDirPath = dest.getAbsolutePath();
        File[] files = input.listFiles();
        if (files.length > 0) {
            for (File file : files) {
                String destFilePath = file.getAbsolutePath().replace(srcDirPath, destDirPath);
                File destFile = new File(destFilePath);
                if (file.isDirectory()) {
                    processDir(file, destFile);
                } else if (file.isFile()) {
                    System.out.println("RouterTransform---processDir---srcDirPath = " + file.getAbsolutePath());
                    System.out.println("RouterTransform---processDir---destDirPath = " + destFile.getAbsolutePath());
//                    weave(file.getAbsolutePath(), destFile.getAbsolutePath());
                    FileUtils.copyFile(file, destFile);
                }
            }
        }
    }

    private void weave(String inputPath, String outputPath) {
        System.out.println("RouterTransform---weave---start---");
        System.out.println("RouterTransform---weave---inputPath = " + inputPath);
        System.out.println("RouterTransform---weave---outputPath = " + outputPath);
        try {
            FileInputStream fileInputStream = new FileInputStream(inputPath);
            ClassReader classReader = new ClassReader(fileInputStream);
            ClassWriter classWriter = new ClassWriter(classReader, ClassWriter.COMPUTE_FRAMES);
            RouterClassVisitor routerClassVisitor = new RouterClassVisitor(Opcodes.ASM7, classWriter, null);
            classReader.accept(routerClassVisitor, ClassReader.EXPAND_FRAMES);
            FileOutputStream fileOutputStream = new FileOutputStream(new File(outputPath));
            fileOutputStream.write(classWriter.toByteArray());
            fileOutputStream.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println("RouterTransform---weave---end---");
    }


    private void processJar(File input, File dest) throws IOException {
        if (input == null || dest == null) {
            return;
        }
        if (dest.exists()) {
            FileUtils.delete(dest);
        }
        FileUtils.mkdirs(dest);
        String srcDirPath = input.getAbsolutePath();
        String destDirPath = dest.getAbsolutePath();
        File[] files = input.listFiles();
        if (files.length > 0) {
            for (File file : files) {
                String destFilePath = file.getAbsolutePath().replace(srcDirPath, destDirPath);
                File destFile = new File(destFilePath);
                if (file.isDirectory()) {
                    processJar(file, destFile);
                } else if (file.isFile()) {
                    // file.getAbsolutePath(), destFile.getAbsolutePath()
                    System.out.println("RouterTransform---processJar---inputPath = " + file.getAbsolutePath());
                    System.out.println("RouterTransform---processJar---outputPath = " + destFile.getAbsolutePath());
                }
            }
        }
    }
}

File: buildSrc/src/main/java/com/mars/router/plugin/RouterTransform.kt
package com.mars.router.plugin

import com.android.build.api.transform.Format
import com.android.build.api.transform.TransformInvocation
import com.android.utils.FileUtils

/**
 * Created by JohnnySwordMan on 2022/1/9
 */
class RouterTransform : BaseTransform() {

    override fun getName(): String {
        return RouterTransform::class.java.simpleName
    }

    override fun transform(transformInvocation: TransformInvocation?) {
        super.transform(transformInvocation)
        transformInvocation ?: return
        // 如果不是增量编译，需要删除
        if (!transformInvocation.isIncremental) {
            transformInvocation.outputProvider.deleteAll()
        }
        // 消费型输入，可以从中获取jar包和class文件夹路径，需要输出给下一个任务
        val inputs = transformInvocation.inputs
        // OutputProvider管理消费型输出，如果消费型输入为空，你会发现OutputProvider为null
        val outputProvider = transformInvocation.outputProvider

        val collector = RouterMappingCollector()

        inputs?.forEach { input ->
            // 本地依赖或者远程依赖，以jar包的形式获取
            input.jarInputs?.forEach {
                val dest = outputProvider?.getContentLocation(
                    it.name,
                    it.contentTypes,
                    it.scopes,
                    Format.JAR
                )
                // 将修改后的字节码copy到dest，就可以实现编译期间干预字节码的目的了
//                collector.collectFromJarFile(it.file)
                FileUtils.copyFile(it.file, dest)

                // 该方法必须在copyFile之后，入参得是dest，
                // 如果放在前面调用，那么找到的Router.class所在的jar包是老的，还没有转换文件路径之前的，因此利用ASM修改，发现一直没有生效
                // 因为这个jar包不会打进apk中
                collector.collectFromJarFile(dest!!)
            }

            // apply该插件的模块的所有的类以directory的形式获取
            input.directoryInputs?.forEach {
                val dest = outputProvider?.getContentLocation(
                    it.name,
                    it.contentTypes,
                    it.scopes,
                    Format.DIRECTORY
                )
//                println("RouterTransform directoryInputs inputFile = ${it.file.absolutePath}")
//                println("RouterTransform directoryInputs outputFile = ${dest?.absolutePath}")

                // 这里需要写copyDirectory，使用copyFile会报异常，例如：RouterApp类不在base.apk中
                collector.collect(it.file)
                FileUtils.copyDirectory(it.file, dest)
            }
        }

        println("RouterTransform all mapping className = ${collector.routerMapping}")

        collector.destFile?.let {
            RegisterCodeGenerator.insertInitCode(collector.routerMapping, it)
        }
        collector.serviceImplFile?.let { file ->
//            RegisterCodeGenerator.insertServiceImplCode(collector.serviceMap, file)
        }
    }
}